### PR TITLE
RIR SSA check pass

### PR DESCRIPTION
**Pull request by swernli**
This pass checks that the RIR is in Single Static Assignment (SSA) form. The check verifies that each variable is assigned exactly once before it is used, and that phi nodes are used correctly.

The change also updates how successors are calculated for branch instructions that have constant conditions, and updates the tests that used constant conditions to use dynamic bool values instead.

### CODE REVIEW

File: compiler/qsc_rir/src/passes/build_dominator_graph/tests.rs [hunk 1]
```diff
@@ -51,7 +37,8 @@ fn test_dominator_graph_single_block_dominates_itself() {
         .insert(BlockId(0), Block(vec![Instruction::Return]));
 
     remap_block_ids(&mut program);
```
Comment (idavis) [inline compiler/qsc_rir/src/passes/build_dominator_graph/tests.rs:39]: Does it make sense to extract theses three calls into a common fn? All of the tests use the same setup.

File: compiler/qsc_rir/src/passes/build_dominator_graph/tests.rs [hunk 2]
```diff
@@ -86,13 +74,38 @@ fn test_dominator_graph_sequential_blocks_dominated_by_predecessor() {
 #[test]
 fn test_dominator_graph_branching_blocks_dominated_by_common_predecessor() {
```
Comment (idavis) [inline compiler/qsc_rir/src/passes/build_dominator_graph/tests.rs:75]: Test names don't really need to have test in the name, they are already attributed with `#[test]`.

File: compiler/qsc_rir/src/passes/ssa_check.rs [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+    doms: &IndexMap<BlockId, BlockId>,
+) {
+    check_phi_nodes(program, preds);
+    let variable_assignments = get_variable_assignments(program);
+    let variable_uses = get_variable_uses(program);
+
+    for (var_id, uses) in variable_uses.iter() {
+        let (def_block_id, def_idx) = variable_assignments
+            .get(var_id)
+            .unwrap_or_else(|| panic!("{var_id:?} is used but not assigned"));
```
Comment (idavis) [inline compiler/qsc_rir/src/passes/ssa_check.rs:29]: Would a let...else block be better here?

File: compiler/qsc_rir/src/passes/ssa_check.rs [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+
+                // Multiple variables
+                Instruction::Call(_, vals, _) => {
+                    for val in vals {
+                        if let Operand::Variable(var) = val {
+                            add_use(var.variable_id, block_id, idx);
+                        }
+                    }
+                }
+                Instruction::Phi(vals, _) => {
```
Comment (idavis) [inline compiler/qsc_rir/src/passes/ssa_check.rs:191]: In other parts of the code base we refer to the `Vec<_>` component of phi nodes as args. We should be consistent one way or another.

File: compiler/qsc_rir/src/passes/ssa_check.rs (hunk 3 above)
Comment (swernli) [reply compiler/qsc_rir/src/passes/ssa_check.rs:29]: Ah, good call. This went through a couple versions, so I didn't notice I had approached but not reached that pattern.

File: compiler/qsc_rir/src/passes/build_dominator_graph/tests.rs (hunk 1 above)
Comment (swernli) [reply compiler/qsc_rir/src/passes/build_dominator_graph/tests.rs:39]: I can do that here and in the other test file as well.

**Review by idavis — APPROVED** [review]

**Pull request merged**